src/test/libslic3r/test_transformationmatrix.cpp

```diff
@@ -0,0 +1,219 @@
+        }
+        WHEN("testing rotation") {
+            THEN("testing axis aligned rotation") {
+                auto mat = TransformationMatrix::mat_rotation(degtorad(90), Axis::X);
+                REQUIRE(check_point(mat.transform(vec1),1,-3,2));
+                mat = TransformationMatrix::mat_rotation(degtorad(90), Axis::Y);
+                REQUIRE(check_point(mat.transform(vec1),3,2,-1));
+                mat = TransformationMatrix::mat_rotation(degtorad(90), Axis::Z);
+                REQUIRE(check_point(mat.transform(vec1),-2,1,3));
+            }
+            THEN("testing arbituary axis rotation") {
+                auto mat = TransformationMatrix::mat_rotation(degtorad(80), vec2);
+                REQUIRE(check_point(mat.transform(vec1),3.0069,1.8341,-1.2627));
+                REQUIRE(std::abs(mat.determinante() - 1.0) < 1.0e-3);
+            }
+            THEN("testing quaternion rotation") {
+                auto mat = TransformationMatrix::mat_rotation(-0.4775,0.3581,-0.2387,0.7660);
+                REQUIRE(check_point(mat.transform(vec1),3.0069,1.8341,-1.2627));
+                REQUIRE(std::abs(mat.determinante() - 1.0) < 1.0e-3);
+            }
+            THEN("testing vector to vector") {
+                auto mat = TransformationMatrix::mat_rotation(vec1,vec2);
+                REQUIRE(check_point(mat.transform(vec1),-2.7792,2.0844,-1.3896));
+                REQUIRE(std::abs(mat.determinante() - 1.0) < 1.0e-3);
+                mat = TransformationMatrix::mat_rotation(vec1,vec1.negative());
+                REQUIRE(check_point(mat.transform(vec1),-1,-2,-3)); // colinear, opposite direction
+                mat = TransformationMatrix::mat_rotation(vec1,vec1);
+                REQUIRE(check_elements(mat, 
+                1.0, 0.0, 0.0, 0.0,
+                0.0, 1.0, 0.0, 0.0,
+                0.0, 0.0, 1.0, 0.0)); // colinear, same direction
+            }
+        }
+    }
+}
+
+bool check_elements(const TransformationMatrix & matrix, 
+    double m00, double m01, double m02, double m03,
+    double m10, double m11, double m12, double m13,
+    double m20, double m21, double m22, double m23)
+{
+    const double EPSILON = 1.0e-3;
+    bool equal = true;
+    equal &= std::abs(matrix.m00 - m00) < EPSILON;
+    equal &= std::abs(matrix.m01 - m01) < EPSILON;
+    equal &= std::abs(matrix.m02 - m02) < EPSILON;
+    equal &= std::abs(matrix.m03 - m03) < EPSILON;
+    equal &= std::abs(matrix.m10 - m10) < EPSILON;
+    equal &= std::abs(matrix.m11 - m11) < EPSILON;
+    equal &= std::abs(matrix.m12 - m12) < EPSILON;
+    equal &= std::abs(matrix.m13 - m13) < EPSILON;
+    equal &= std::abs(matrix.m20 - m20) < EPSILON;
+    equal &= std::abs(matrix.m21 - m21) < EPSILON;
+    equal &= std::abs(matrix.m22 - m22) < EPSILON;
+    equal &= std::abs(matrix.m23 - m23) < EPSILON;
+    return equal;
+}
+
+bool check_point(const Pointf3 & point, coordf_t x, coordf_t y, coordf_t z)
+{
+    const coordf_t EPSILON = 1.0e-3;
```

Comment:
Does this fail using libslic3r's epsilon constant?

Reply:
Probably not, I think I relaxed that threshold as it compares to the manual values given, not to the numerical precision. Bad naming from me, rename to EQUALITY_THRESHOLD (defined in the header) or adapt to the EPSILON (whose comments suggest only being used for floating-point precision stuff)?

Reply:
I pushed the threshold definition version